src/Dev/CsvBulkLoader.php

```diff
@@ -67,37 +67,74 @@ public function preview($filepath)
+        $previousDetectLE = ini_get('auto_detect_line_endings');
 
+        ini_set('auto_detect_line_endings', true);
         try {
-            foreach ($files as $file) {
-                $last = $file;
+            $filepath = Director::getAbsFile($filepath);
+            $csvReader = Reader::createFromPath($filepath, 'r');
+
+            if ($this->columnMap) {
+                $headerMap = $this->getNormalisedColumnMap();
+                $remapper = function ($row, $rowOffset, $iterator) use ($headerMap) {
+                    foreach ($headerMap as $column => $renamedColumn) {
+                        if ($column == $renamedColumn) {
+                            continue;
+                        }
+                        if (array_key_exists($column, $row)) {
+                            if (strpos($renamedColumn, '_ignore_') !== 0) {
+                                $row[$renamedColumn] = $row[$column];
+                            }
+                            unset($row[$column]);
+                        }
+                    }
+                    return $row;
+                };
+            } else {
+                $remapper = null;
+            }
 
-                $next = $this->processChunk($file, $preview);
+            if ($this->hasHeaderRow) {
+                $rows = $csvReader->fetchAssoc(0, $remapper);
+            } elseif ($this->columnMap) {
+                $rows = $csvReader->fetchAssoc($headerMap, $remapper);
+            }
 
-                if ($result instanceof BulkLoader_Result) {
-                    $result->merge($next);
-                } else {
-                    $result = $next;
-                }
+            $result = BulkLoader_Result::create();
 
-                @unlink($file);
+            foreach ($rows as $row) {
+                $this->processRecord($row, $this->columnMap, $result, $preview);
             }
-        } catch (Exception $e) {
-            $failedMessage = sprintf("Failed to parse %s", $last);
+        } catch (\Exception $e) {
+            $failedMessage = sprintf("Failed to parse %s", $filepath);
             if (Director::isDev()) {
                 $failedMessage = sprintf($failedMessage . " because %s", $e->getMessage());
             }
             print $failedMessage . PHP_EOL;
+        } finally {
+            ini_set('auto_detect_line_endings', $previousDetectLE);
         }
-
         return $result;
     }
 
+    protected function getNormalisedColumnMap()
+    {
+        $map = [];
+        foreach ($this->columnMap as $column => $newColumn) {
+            if (strpos($newColumn, "->") === 0) {
+                $map[$column] = $column;
+            } elseif (is_null($newColumn)) {
+                // the column map must consist of unique scalar values
+                // `null` can be present multiple times and is not scalar
+                // so we name it in a standard way so we can remove it later
+                $map[$column] = '_ignore_' . $column;
```

Comment:
Looks like this is new. If `null` isn’t supported, should it raise a notice of some sort?

Reply:
this is for compatibility with `league/csv`, rather than changing the API for SS users.

This keeps the existing SS API / behaviour.